src/protocols/secure_channel/StatusReport.cpp

```diff
@@ -0,0 +1,73 @@
+/**
+ *    @file
+ *      Implements methods for the StatusReport class.
+ */
+
+#include <protocols/secure_channel/StatusReport.h>
+
+#include <support/BufferReader.h>
+#include <support/CodeUtils.h>
+#include <support/ReturnMacros.h>
+
+using namespace chip::Encoding;
+
+namespace chip {
+namespace Protocols {
+namespace SecureChannel {
+
+StatusReport::StatusReport() : mGeneralCode(0), mProtocolId(0), mProtocolCode(0), mProtocolData(nullptr), mProtocolDataLength(0) {}
+
+StatusReport::StatusReport(uint16_t generalCode, uint32_t protocolId, uint16_t protocolCode) :
+    mGeneralCode(generalCode), mProtocolId(protocolId), mProtocolCode(protocolCode), mProtocolData(nullptr), mProtocolDataLength(0)
+{}
+
+StatusReport::StatusReport(uint16_t generalCode, uint32_t protocolId, uint16_t protocolCode, uint8_t * protocolData,
+                           uint16_t dataLen) :
+    mGeneralCode(generalCode),
+    mProtocolId(protocolId), mProtocolCode(protocolCode), mProtocolData(protocolData), mProtocolDataLength(dataLen)
+{}
+
+CHIP_ERROR StatusReport::Parse(System::PacketBufferHandle buf)
+{
+    ReturnErrorCodeIf(buf.IsNull(), CHIP_ERROR_INVALID_ARGUMENT);
+
+    uint8_t * bufStart = buf->Start();
+    LittleEndian::Reader bufReader(bufStart, buf->DataLength());
+
+    ReturnErrorOnFailure(bufReader.Read16(&mGeneralCode).Read32(&mProtocolId).Read16(&mProtocolCode).StatusCode());
+
+    if (bufReader.OctetsRead() < buf->DataLength())
```

Comment:
Suggest commenting that if leftover data exists after header, it is error-specific data and ownership of the underlying buffer is taken.

When does the mMsgHandle get released?

Reply:
Will add comment.

mMsgHandle will be released when the StatusReport goes out of scope. PacketBufferHandles have a destructor that releases the PacketBuffer